File: Chauffeur/Deliverables/UserDeliverable.cs
using System;
using System.IO;
using System.Linq;
using System.Threading.Tasks;
using umbraco;
using Umbraco.Core.Logging;
using Umbraco.Core.Models.Identity;
using Umbraco.Core.Models.Membership;
using Umbraco.Core.Security;
using Umbraco.Core.Services;

namespace Chauffeur.Deliverables
{
    [DeliverableName("user")]
    [DeliverableAlias("u")]
    public sealed class UserDeliverable : Deliverable, IProvideDirections
    {
        private readonly IUserService userService;
        private readonly BackOfficeUserManager<BackOfficeIdentityUser> userManager;

        public UserDeliverable(
            TextReader reader,
            TextWriter writer,
            IUserService userService,
            BackOfficeUserManager<BackOfficeIdentityUser> userManager
        ) : base(reader, writer)
        {
            this.userService = userService;
            this.userManager = userManager;
        }

        public override async Task<DeliverableResponse> Run(string command, string[] args)
        {
            if (!args.Any())
            {
                await Out.WriteLineAsync("No operation for the user was provided");
                return DeliverableResponse.Continue;
            }

            var operation = args[0];

            switch (operation)
            {
                case "change-password":
                    await ChangePassword(args.Skip(1).ToArray());
                    return DeliverableResponse.Continue;

                case "change-name":
                    await ChangeName(args.Skip(1).ToArray());
                    return DeliverableResponse.Continue;

                case "change-loginname":
                    await ChangeLoginName(args.Skip(1).ToArray());
                    return DeliverableResponse.Continue;

                case "create-user":
                    return await CreateUser(args.Skip(1).ToArray());

                default:
                    await Out.WriteLineFormattedAsync("The user operation '{0}' is not supported", operation);
                    return DeliverableResponse.Continue;
            }
        }

        private async Task<DeliverableResponse> CreateUser(string[] args)
        {
            if (args.Length != 5)
            {
                await Out.WriteLineAsync("Please provide 5 arguments, name, username, email, password and groups. For more information see `help`");
                return DeliverableResponse.Continue;
            }

            var name = args[0];
            var username = args[1];
            var email = args[2];
            var password = args[3];
            var groupNames = args[4];

            var identity = BackOfficeIdentityUser.CreateNew(username, email, GlobalSettings.DefaultUILanguage);
            identity.Name = name;

            var result = await userManager.CreateAsync(identity);

            if (!result.Succeeded)
            {
                await Out.WriteLineAsync("Error saving the user:");
                foreach (var error in result.Errors)
                    await Out.WriteLineAsync($"\t{error}");

                return DeliverableResponse.FinishedWithError;
            }

            result = await userManager.AddPasswordAsync(identity.Id, password);
            if (!result.Succeeded)
            {
                await Out.WriteLineAsync("Error saving the user password:");
                foreach (var error in result.Errors)
                    await Out.WriteLineAsync($"\t{error}");

                return DeliverableResponse.FinishedWithError;
            }

            var user = userService.GetByEmail(email);
            var groups = userService.GetUserGroupsByAlias(groupNames.Split(new[] { ',' }, StringSplitOptions.RemoveEmptyEntries));
            foreach (var group in groups)
            {
                var rg = new ReadOnlyUserGroup(group.Id, group.Name, group.Icon, group.StartContentId, group.StartMediaId, group.Alias, group.AllowedSections, group.Permissions);
                user.AddGroup(rg);
            }
            user.IsApproved = true;
            userService.Save(user);

            return DeliverableResponse.Continue;
        }

        private async Task ChangePassword(string[] args)
        {
            if (args.Length != 2)
            {
                await Out.WriteLineAsync("The expected parameters for 'change-password' were not supplied.");
                await Out.WriteLineAsync("Format expected: change-password <username> <new password>");
                return;
            }

            var username = args[0];
            var user = userService.GetByUsername(username);

            if (user == null)
            {
                await Out.WriteLineFormattedAsync("User '{0}' does not exist in the system", username);
                return;
            }

            try
            {
                userService.SavePassword(user, args[1]);
                await Out.WriteLineFormattedAsync("User '{0}' has had their password updated", username);
            }
            catch (NotSupportedException ex)
            {
                LogHelper.Error<UserDeliverable>("Wasn't able to update the user password from Chauffeur", ex);

                await Out.WriteLineAsync("Updating the user password is not supported.");
                await Out.WriteLineAsync("It's most likely because your UsersMembershipProvider has 'allowManuallyChangingPassword=\"false\"'.");
                await Out.WriteLineAsync("Currently Chauffeur can't update passwords for membership providers configured like this.");
            }
        }

        private async Task ChangeName(string[] args)
        {
            if (args.Length != 2)
            {
                await Out.WriteLineAsync("The expected parameters for 'change-name' were not supplied.");
                await Out.WriteLineAsync("Format expected: change-name <username> <new username>");
                return;
            }

            var username = args[0];
            var user = userService.GetByUsername(username);

            if (user == null)
            {
                await Out.WriteLineFormattedAsync("User '{0}' does not exist in the system", username);
                return;
            }
            user.Name = args[1];
            userService.Save(user);
            await Out.WriteLineFormattedAsync("User '{0}' has had their name updated", username);
        }

        private async Task ChangeLoginName(string[] args)
        {
            if (args.Length != 2)
            {
                await Out.WriteLineAsync("The expected parameters for 'change-loginname' were not supplied.");
                await Out.WriteLineAsync("Format expected: change-loginname <username> <new loginname>");
                return;
            }

            var username = args[0];
            var user = userService.GetByUsername(username);

            if (user == null)
            {
                await Out.WriteLineFormattedAsync("User '{0}' does not exist in the system", username);
                return;
            }
            user.Username = args[1];
            userService.Save(user);
            await Out.WriteLineFormattedAsync("User '{0}' has had their login name updated", username);
        }

        public async Task Directions()
        {
            await Out.WriteLineAsync("A series of operations that can be run against an Umbraco User.");
            await Out.WriteLineAsync();

            await Out.WriteLineAsync("change-password <username> <new password>");
            await Out.WriteLineAsync("\tChanges the password for a given user. This will also hash it if hashing is turned on in the web.config");
            await Out.WriteLineAsync("change-name <username> <new username>");
            await Out.WriteLineAsync("\tChanges the user name for a given user.");
            await Out.WriteLineAsync("change-loginname <username> <new loginname>");
            await Out.WriteLineAsync("\tChanges the login name for a given user.");
        }
    }
}

File: src/Chauffeur/Deliverables/UserDeliverable.fs
namespace Chauffeur.Deliverables

open Chauffeur
open FSharp.Control.Tasks.V2
open Umbraco.Core.Services
open Umbraco.Core.Configuration
open Umbraco.Web.Security
open Umbraco.Core.Models.Identity
open Umbraco.Core.Logging
open Umbraco.Core.Models.Membership
open System.Threading.Tasks
open System

module UserPasswordChanger =
    let resetPasswordWithToken generateToken (resetPassword : (int * string * string) -> Task<Microsoft.AspNet.Identity.IdentityResult>) userId password =
        task {
            let! token = generateToken userId
            let! result = resetPassword (userId, token, password)

            return result
        }

    let password write (getUser : string -> IUser) (resetPassword : int -> string -> Task<Microsoft.AspNet.Identity.IdentityResult>) args =
        task {
            match args with
            | username :: password :: _ ->
                let user = getUser username
                match user with
                | null ->
                    do! (sprintf "User '%s' does not exist in the system" username) |> write
                | _ ->
                    let! result = resetPassword user.Id password
                    match result.Succeeded with
                    | true ->
                        do! (sprintf "User '%s' has had their password changed" username) |> write
                    | false ->
                        do! write "There were errors changing the password:"
                        for err in result.Errors do
                            do! (sprintf "\t%s" err) |> write
            | _ ->
                do! write("The expected parameters for 'change-password' were not supplied.");
                do! write("Format expected: change-password <username> <new password>");

            return ignore()
        }

module UserDetailsChanger =
    let name writer (getUser : string -> IUser) saveUser args =
        task {
            match args with
            | username :: name :: _ ->
                let user = getUser username
                match user with
                | null ->
                    do! (sprintf "User '%s' does not exist in the system" username) |> writer
                | _ ->
                    user.Name <- name
                    saveUser user
                    do! (sprintf "User '%s' has had their name updated" username) |> writer

            | _ ->
                do! writer "The expected parameters for 'change-name' were not supplied."
                do! writer "Format expected: change-name <username> <new username>"
        }

    let loginName writer (getUser : string -> IUser) saveUser  args =
        task {
            match args with
            | username :: newUsername :: _ ->
                let user = getUser username
                match user with
                | null ->
                    do! (sprintf "User '%s' does not exist in the system" username) |> writer
                | _ ->
                    user.Username <- newUsername
                    saveUser user
                    do! (sprintf "User '%s' has had their name updated" username) |> writer

            | _ ->
                do! writer "The expected parameters for 'change-loginname' were not supplied."
                do! writer "Format expected: change-loginname <username> <new username>"
        }

module UserCreator =
    let createUser (writer : string -> Task) uiLanguage (create : BackOfficeIdentityUser -> Task<Microsoft.AspNet.Identity.IdentityResult>) name username email =
        task {
            let identity = BackOfficeIdentityUser.CreateNew(username, email, uiLanguage)
            identity.Name <- name
            
            let! result = create identity

            match result.Succeeded with
            | false ->
                do! writer "Error saving the user:"
                for err in result.Errors do
                    do! writer(sprintf "\t%s" err)
                return None
            | true ->
                return Some identity
        }

    let addPassword (writer : string -> Task) (addPassword : int * string -> Task<Microsoft.AspNet.Identity.IdentityResult>) password (identity : BackOfficeIdentityUser option) =
        task {
            match identity with
            | None -> return None
                | Some identity-> 

                let! result = addPassword(identity.Id, password)
                match result.Succeeded with
                | false ->
                    do! writer "Error saving the user:"
                    for err in result.Errors do
                        do! writer(sprintf "\t%s" err)
                    return None
                | true ->
                    return Some identity
        }

    let addGroups (getUser : string -> IUser) (groupsByAlias : string array -> IUserGroup seq) saveUser groups (identity : BackOfficeIdentityUser option) =
        match identity with
        | None -> None
        | Some identity-> 
            let user = getUser identity.Email
            let userGroups = groupsByAlias groups
            for g in userGroups do
                let rg = ReadOnlyUserGroup(g.Id, g.Name, g.Icon, g.StartContentId, g.StartMediaId, g.Alias, g.AllowedSections, g.Permissions)
                user.AddGroup rg

            user.IsApproved <- true
            saveUser user
            Some user

    let create writer (createUser : string -> string-> string -> Task<BackOfficeIdentityUser option>) addPassword addGroups args =
        task {
            match args with
            | name :: username :: email :: password :: groupNames :: _ ->
                let! identity = createUser name username email
                let! _ = addPassword password identity
                let user = addGroups (groupNames.Split([| ',' |], StringSplitOptions.RemoveEmptyEntries)) identity

                match user with
                | Some _ ->
                    do! (sprintf "The user '%s' has been created" username) |> writer
                | None ->
                    ignore()

            | _ ->
                do! writer "Please provide 5 arguments, name, username, email, password and groups. For more information see `help`"
        }

[<DeliverableName("user")>]
[<DeliverableAlias("u")>]
type UserDeliverable(
                    reader,
                    writer,
                    userService : IUserService,
                    userManager : BackOfficeUserManager<BackOfficeIdentityUser>,
                    umbracoSettings : IGlobalSettings) =
    inherit Deliverable(reader, writer)

    let changePassword = UserPasswordChanger.password writer.WriteLineAsync
    let changeName = UserDetailsChanger.name writer.WriteLineAsync
    let changeLoginName = UserDetailsChanger.loginName writer.WriteLineAsync

    override __.Run _ args =
        task {
            match args |> Array.toList with
            | [] ->
                do! writer.WriteLineAsync "No operation for the user was provided"
            | head :: tail when head = "change-password" ->
                let passwordResetter = 
                    UserPasswordChanger.resetPasswordWithToken (userManager.GeneratePasswordResetTokenAsync) (userManager.ResetPasswordAsync)

                do! changePassword (userService.GetByUsername) passwordResetter tail
            | head :: tail when head = "change-name" ->
                do! changeName (userService.GetByUsername) (userService.Save) tail
            | head :: tail when head = "change-loginname" ->
                do! changeLoginName (userService.GetByUsername) (userService.Save) tail
            | head :: tail when head = "create-user" ->
                let createUser = UserCreator.createUser (writer.WriteLineAsync) umbracoSettings.DefaultUILanguage (userManager.CreateAsync)
                let addPassword = UserCreator.addPassword (writer.WriteLineAsync) (userManager.AddPasswordAsync)
                let addGroups = UserCreator.addGroups (userService.GetByEmail) (fun g -> userService.GetUserGroupsByAlias g) (userService.Save)

                do! UserCreator.create (writer.WriteLineAsync) createUser addPassword addGroups tail
            | _ ->
                do! writer.WriteLineAsync(sprintf "The user operation '%s' is not supported" args.[0])

            return DeliverableResponse.Continue
        }